tests/unit/test_table.py

```diff
@@ -857,6 +851,74 @@ def test_read_retry_rows(self):
         result = rows[1]
         self.assertEqual(result.row_key, self.ROW_KEY_2)
 
+    def test_read_retry_rows_timeouts(self):
+        from google.cloud.bigtable_v2.gapic import bigtable_client
+
+        data_api = bigtable_client.BigtableClient(mock.Mock())
+        credentials = _make_credentials()
+        client = self._make_client(project="project-id", credentials=credentials)
+        client._table_data_client = data_api
+        instance = client.instance(instance_id=self.INSTANCE_ID)
+        table = self._make_one(self.TABLE_ID, instance)
+
+        # Patch the stub used by the API method.
+        client._table_data_client.transport.read_rows = mock.Mock(
+            side_effect=[_MockReadRowsIterator()]
+        )
+
+        # By default there is no timeout
+        list(table.read_rows())
+        self.assertIsNone(
+            client._table_data_client.transport.read_rows.call_args.kwargs["timeout"]
+        )
+
+        # attempt timeout should be passed thru
+        client._table_data_client.transport.read_rows = mock.Mock(
+            side_effect=[_MockReadRowsIterator()]
+        )
+        list(table.read_rows(attempt_timeout=1.0))
+        self.assertEquals(
+            1.0,
+            client._table_data_client.transport.read_rows.call_args.kwargs["timeout"],
+        )
+
+        # overall timeout should be passed thru
+        client._table_data_client.transport.read_rows = mock.Mock(
+            side_effect=[_MockReadRowsIterator()]
+        )
+        list(table.read_rows(overall_timeout=10.0))
+        self.assertLess(
+            8.0,
+            client._table_data_client.transport.read_rows.call_args.kwargs["timeout"],
+        )
+
+        # attempt timeout limits overall timeout
+        client._table_data_client.transport.read_rows = mock.Mock(
+            side_effect=[_MockReadRowsIterator()]
+        )
+        list(table.read_rows(attempt_timeout=5.0, overall_timeout=10.0))
+        self.assertLessEqual(
+            5.0,
+            client._table_data_client.transport.read_rows.call_args.kwargs["timeout"],
+        )
+
+        # attempt timeout is truncated by overall timeout
+        class DelayedFailureIterator(object):
+            def next(self):
+                time.sleep(0.75)
+                raise DeadlineExceeded("delayed error")
+
+            __next__ = next
+
+        client._table_data_client.transport.read_rows = mock.Mock(
+            side_effect=[DelayedFailureIterator(), _MockReadRowsIterator()]
+        )
+        list(table.read_rows(attempt_timeout=1.0, overall_timeout=1.0))
+        self.assertGreater(
```

Comment:
Hmm why is this greater?

Reply:
are you trying to test where the attempt_timetout > overall_timeout after a few runs?

Reply:
No, the assert api is terrible
self.assertGreater(a,b) means

if not a > b:
  fail

so this is asserting that the rpc timeout is less than the attempt timeout